### PR TITLE
Fix ruin spawning, LZ deletion, UTs and more

### PR DESCRIPTION
**Pull request by JohnWildkins**
Remaining issues (with the ruin system, out of scope):
  - TEMPLATE_FLAG_CLEAR_CONTENTS does nothing
  - ~~TEMPLATE_FLAG_SPAWN_GUARANTEED does not work on ruins, only away sites~~
  - no way to easily jump to ruins without inputting coordinate from log

changes:
  - bugfix: "Fixed an issue that prevented several ruins from spawning on exoplanets."
  - bugfix: "Fixed ruin spawns being partially / wholly deleted by landing zones, causing unit test failures and, of course, deleted ruins."
  - backend: "Changed the ruin tag system to more appropriately reflect types that actually want to be filtered for. Added planetary tags to restrict ruins by planet type."
  - backend: "The away site testing map now attempts to build one of each exoplanet (excluding lore planets) and one of each ruin."

### CODE REVIEW

**Comment by JohnWildkins** [comment]
!review

File: code/__defines/misc.dm [hunk 1]
```diff
@@ -497,13 +497,13 @@ Define for getting a bitfield of adjacent turfs that meet a condition.
 #define TEMPLATE_FLAG_NO_RUINS         8 // if it should forbid ruins from spawning on top of it
 
 //Ruin map template flags
-#define TEMPLATE_FLAG_RUIN_STARTS_DISALLOWED 32  // Ruin is not available during spawning unless another ruin permits it.
+#define TEMPLATE_FLAG_RUIN_STARTS_DISALLOWED 32  // Ruin is not available during spawning unless another ruin permits it, or whitelisted by the exoplanet
```
Comment (SleepyGemmy) [inline code/__defines/misc.dm:500]:
```suggestion
#define TEMPLATE_FLAG_RUIN_STARTS_DISALLOWED 32  // Ruin is not available during spawning unless another ruin permits it, or is whitelisted by the exoplanet.
```

File: code/controllers/subsystems/initialization/xenoarch.dm [hunk 2]
```diff
@@ -18,6 +18,7 @@ var/datum/controller/subsystem/xenoarch/SSxenoarch
 	NEW_SS_GLOBAL(SSxenoarch)
 
 /datum/controller/subsystem/xenoarch/Initialize(timeofday)
+	set background = 1
```
Comment (SleepyGemmy) [inline code/controllers/subsystems/initialization/xenoarch.dm:21]:
```suggestion
```
undocumented and out of scope change.

File: code/game/objects/random/random.dm [hunk 3]
```diff
@@ -1227,6 +1227,24 @@
...
+		/obj/item/clothing/head/hardhat/red = 1,
+		/obj/item/clothing/head/hardhat/green = 1,
+		/obj/item/clothing/head/hardhat/dblue = 1,
+		/obj/item/clothing/head/hardhat/white = 0.5,
+		/obj/item/clothing/head/hardhat/atmos = 0.1,
+		/obj/item/clothing/head/hardhat/first_responder = 0.1,
+		/obj/item/clothing/head/hardhat/firefighter = 0.1
+	)
+
+
```
Comment (SleepyGemmy) [inline code/game/objects/random/random.dm:1247]:
```suggestion
```

File: code/modules/maps/ruins.dm [hunk 4]
```diff
@@ -13,55 +13,50 @@ var/list/banned_ruin_ids = list()
 
 	var/list/available = list()
 	var/list/selected = list()
+	var/list/force_spawn = list()
 	var/remaining = budget
 
 	for(var/datum/map_template/ruin/ruin in potentialRuins)
+		if (ruin.template_flags & TEMPLATE_FLAG_SPAWN_GUARANTEED)
```
Comment (SleepyGemmy) [inline code/modules/maps/ruins.dm:20]:
```suggestion
		if(ruin.template_flags & TEMPLATE_FLAG_SPAWN_GUARANTEED)
```

File: code/modules/maps/ruins.dm [hunk 5]
```diff
@@ -13,55 +13,50 @@ var/list/banned_ruin_ids = list()
...
+			force_spawn |= ruin
+			continue
 		if (ruin.id in banned_ruin_ids)
```
Comment (SleepyGemmy) [inline code/modules/maps/ruins.dm:23]:
```suggestion
		if(ruin.id in banned_ruin_ids)
```

File: code/modules/maps/ruins.dm [hunk 6]
```diff
@@ -13,55 +13,50 @@ var/list/banned_ruin_ids = list()
...
 			continue
-		if(!(SSatlas.current_sector.name in ruin.sectors))
+		if(!(SSatlas.current_sector.name in ruin.sectors) && !ignore_sector)
 			continue
 		available[ruin] = ruin.spawn_weight
 
-	if (!length(available))
+	if (!length(available) && !length(force_spawn))
```
Comment (SleepyGemmy) [inline code/modules/maps/ruins.dm:29]:
```suggestion
	if(!length(available) && !length(force_spawn))
```

File: code/modules/maps/ruins.dm [hunk 7]
```diff
@@ -13,55 +13,50 @@ var/list/banned_ruin_ids = list()
...
-			selected += ruin
-			if (ruin.spawn_cost > 0)
-				remaining -= ruin.spawn_cost
-			if (!(ruin.template_flags & TEMPLATE_FLAG_ALLOW_DUPLICATES))
-				banned_ruin_ids += ruin.id
-				available -= ruin
-			break
+		if (ruin.spawn_cost > 0)
+			remaining -= ruin.spawn_cost
+		if (!(ruin.template_flags & TEMPLATE_FLAG_ALLOW_DUPLICATES))
```
Comment (SleepyGemmy) [inline code/modules/maps/ruins.dm:57]:
```suggestion
		if(ruin.spawn_cost > 0)
			remaining -= ruin.spawn_cost
		if(!(ruin.template_flags & TEMPLATE_FLAG_ALLOW_DUPLICATES))
```

File: code/modules/maps/ruins.dm [hunk 8]
```diff
@@ -71,6 +66,35 @@ var/list/banned_ruin_ids = list()
...
+
+	var/width = TRANSITIONEDGE + RUIN_MAP_EDGE_PAD + round(ruin.width / 2)
+	var/height = TRANSITIONEDGE + RUIN_MAP_EDGE_PAD + round(ruin.height / 2)
+
+	if(width > max_x - width || height > max_y - height)
+		return
+
+	var/valid = TRUE
+
+	for (var/attempts = 20, attempts > 0, --attempts)
```
Comment (SleepyGemmy) [inline code/modules/maps/ruins.dm:84]:
```suggestion
	for(var/attempts = 20, attempts > 0, --attempts)
```

**Review by NonQueueingMatt — APPROVED** [review]

**Review by Alberyk — APPROVED** [review]

**Pull request merged**